src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -600,6 +611,155 @@ fromChunks cs FinalChunk{..} = CompactIndex {
     }
 
+{-------------------------------------------------------------------------------
+  Serialisation
+-------------------------------------------------------------------------------}
+
+-- | 32 bit aligned.
+chunkBuilder :: Chunk -> BB.Builder
+chunkBuilder Chunk {..} = VU.foldMap BB.word32LE cPrimary
+
+-- | Must be written after the sequence of 'chunkBuilder' of the corresponding
+-- compact index. Specifically, the written chunks must match 'fcNumPages' to
+-- get the alignment right.
+finalChunkBuilder :: NumEntries -> FinalChunk -> BB.Builder
+finalChunkBuilder (NumEntries numEntries) FinalChunk {..} =
+       VU.foldMap BB.word32LE fcRangeFinder
+    <> (if odd (fcNumPages + VU.length fcRangeFinder)  -- align to 64 bit
+        then BB.word32LE 0
+        else mempty)
+    <> putBitVec fcClashes
+    <> putBitVec fcLargerThanPage
+    <> putTieBreaker fcTieBreaker
+    <> BB.word64LE (fromIntegral fcRangeFinderPrecision)
+    <> BB.word64LE (fromIntegral fcNumPages)
+    <> BB.word64LE (fromIntegral numEntries)
+
+-- | Padded to 64 bit.
+--
+-- Assumes that the bitvector is byte-aligned.
+putBitVec :: VU.Vector Bit -> BB.Builder
+putBitVec (BitVec offsetBits lenBits ba)
+  | mod8 offsetBits /= 0 = error "putBitVec: not byte aligned"
+  | otherwise =
+       -- first only write the bytes that are fully part of the bit vec
+       byteArrayFromTo offsetBytes offsetLastByte ba
+       -- then carefully write the last byte, might be partially uninitialised
+    <> (if remainingBits /= 0 then
+         BB.word8 (indexByteArray ba offsetLastByte .&. bitmaskLastByte)
+       else
+         mempty)
+    <> putPaddingTo64 totalBytesWritten
+  where
+    offsetBytes = div8 offsetBits
+    offsetLastByte = offsetBytes + div8 lenBits
+    totalBytesWritten = divUp8 lenBits
+
+    bitmaskLastByte = unsafeShiftR 0xFF (8 - remainingBits)
+    remainingBits = mod8 lenBits
+
+-- | Padded to 64 bit.
+putTieBreaker :: Map SerialisedKey PageNr -> BB.Builder
+putTieBreaker m =
+       BB.word64LE (fromIntegral (Map.size m))
+    <> foldMap putEntry (Map.assocs m)
+  where
+    putEntry :: (SerialisedKey, PageNr) -> BB.Builder
+    putEntry (k, PageNr pageNr) =
+           BB.word32LE (fromIntegral pageNr)
+        <> BB.word32LE (fromIntegral (sizeofKey k))
+        <> serialisedKey k
+        <> putPaddingTo64 (sizeofKey k)
+
+putPaddingTo64 :: Int -> BB.Builder
+putPaddingTo64 written
+  | mod8 written == 0 = mempty
+  | otherwise         = foldMap BB.word8 (replicate (8 - mod8 written) 0)
+
+{-------------------------------------------------------------------------------
+  Deserialisation
+-------------------------------------------------------------------------------}
+
+-- | The input bytestring must be 64 bit aligned and exactly contain the
+-- serialised compact index, with no leading or trailing space.
+-- It is directly used as the backing memory for the compact index.
+--
+-- Also note that the implementation assumes a little-endian system.
```

Comment:
It's perhaps worth noting that we _could_ support big endian systems (though this impl doesn't currently) by byte swapping all the elements of all the arrays.

We should probably include the host endian check in any the deserialisation sanity checks.

Alternatively, we could allow the format to be little or big endian, and use the range-finder field to detect which endian it is (since that field only uses 5 out of 64 bits). In that case a sanity check that the range finder bits is in the range 0..16 would also cover the endianness check.